---
title: "Class 11: Structural Bioinfomatics pt2"
author: "Peyton Chiu (PID:A18145937)"
format: html
---

## Alphafold Data Base (AFDB)

The EBI maintains the largest database of Alphafold structure prediction models at https://alphafold.ebi.ac.uk/

From last class we saw that the PDB had 244290
(Oct 2025)

The total number of protein seqeuences in UniProtKB is 199,579,901

>Key point: Ths is a tiny fraction of sequence space that has structural coverage

```{r}
244290/199579901 * 100
```

AFDB is attempting to address this gap 

There are two quality scores from Alphafold, one for residues (amino acids) called pLDDT distribution. The other PAE score measures the confidence in the relative position of the two residues(a score for every pair of residues)

## Generating your own structure predictions 

Figure of 5 general HIV-PR  models 
![](s1.png)

![](s2.png)

![](s3.png)
![](S4.png)
![](S5.png) 

And the top ranked model 

![](Top model.png)


pLDDT score for model 1 
![](ppdltTop.png)



and model 5

![](ppdltBot.png)



# Custom Analysis of resulting models in R 

Read key results files into R. The first thing I need to know is what my results directory folder is called 


```{r}

results.dir <- "HIVPR_dimer_23119.result"

# File names for all PDB models
pdb_files <- list.files(path=results.dir,
                        pattern="*.pdb",
                        full.names = TRUE)

# Print our PDB file names
basename(pdb_files)

library(bio3d)

# Read all data from Models 
#  and superpose/fit coords
pdbs <- pdbaln(pdb_files, fit=TRUE, exefile="msa")
```


## RMSD heat map 

```{r}
rd <- rmsd(pdbs, fit=T)
range(rd)
library(pheatmap)

colnames(rd) <- paste0("m",1:5)
rownames(rd) <- paste0("m",1:5)
pheatmap(rd)
```






## Residue Conservation from alignment file 

```{r}
aln_file <- list.files(path=results.dir,
                       pattern=".a3m$",
                        full.names = TRUE)
aln_file
```

```{r}
aln <- read.fasta(aln_file[1], to.upper = TRUE)
```

How many sequences in this file 
```{r}
dim(aln$ali)
```

Scoring residue conservation 

```{r}
sim <- conserv(aln)


```

```{r}

pdb <- read.pdb("1hsg")
plotb3(sim[1:99], sse=trim.pdb(pdb, chain="A"),
       ylab="Conservation Score")
```

```{r}
con <- consensus(aln, cutoff = 0.9)
con$seq
```

